processing/src/main/java/org/apache/druid/timeline/partition/PartitionHolder.java

```diff
@@ -60,14 +67,24 @@ public PartitionHolder(List<PartitionChunk<T>> initialChunks)
     }
   }
 
-  protected PartitionHolder(OvershadowableManager<T> overshadowableManager)
+  protected PartitionHolder(OvershadowableManager<T> overshadowableManager, short maxMinorVersion)
   {
     this.overshadowableManager = overshadowableManager;
+    this.maxMinorVersion = maxMinorVersion;
   }
 
   public boolean add(PartitionChunk<T> chunk)
   {
-    return overshadowableManager.addChunk(chunk);
+    boolean added = overshadowableManager.addChunk(chunk);
```

Comment:
If we do, we would have to maintain counts and complicate the logic.
This patch is intended to help large clusters without segment locking i.e distinct minor versions.
Would it be better if a more elaborate solution which updates the state when remove is called is implemented?

Reply:
Yes, we would have to maintain counts for each minor version.

So if we skip the impl on `remove`, if a minor version was ever added to the timeline, we will continue to assume that that minor version is still present for the purposes of computing maxMinorVersion. So we might sometimes enter into the slow code flow even though it could have been avoided. I guess that's not so bad.

For the timeline in the `DataSourcesSnapshot`, the impl in `remove` doesn't matter (at least in the current implementation of the `SqlSegmentsMetadataManager`) since objects are never removed, only added to the timeline.

I guess we can keep the impl simple for now, as you suggest, but we should atleast call it out in a comment.

Also, since the logic of the `maxMinorVersion` is very adhoc and confusing, we should add comments as appropriate and keep the changes limited to the `VersionedIntervalTimeline` class.